staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go

```diff
@@ -238,11 +247,27 @@ func (r *RequestInfoFactory) NewRequestInfo(req *http.Request) (*RequestInfo, er
 			}
 		}
 	}
+
 	// if there's no name on the request and we thought it was a delete before, then the actual verb is deletecollection
 	if len(requestInfo.Name) == 0 && requestInfo.Verb == "delete" {
 		requestInfo.Verb = "deletecollection"
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.AuthorizeWithSelectors) {
+		switch {
+		case requestInfo.Verb == "list" || requestInfo.Verb == "watch":
```

Comment:
Because `AttributesRecord` is the `authorization.Attributes` interface (not the `*AttributesRecord`), it's actually a pain to make it efficient and thread-safe due to copying the lock by value.

Thoughts on actually doing the parse once here, paying that cost and attaching the parsed result and/or parseError so the `AttributesRecord` remains copyable?  We might parse more than necessary, but we don't have to start considering every stack to determine if we're shallow-copy + mutating today.

Reply:
I think I'd start with the parse in the authz layer like you did above... not sure about parsing here yet

Reply:
> I think I'd start with the parse in the authz layer like you did above... not sure about parsing here yet

So that I wasn't able to do. At least not quickly.  Taking this from a value-struct to a stateful struct that cannot be copied means that the ripples are pretty significant.  We have to update everyone expecting the value-struct to use a different construct or we find some other layer to work in.  Both require refactoring.